src/fsharp/Optimizer.fs

```diff
@@ -1773,9 +1780,54 @@ and OptimizeInterfaceImpl cenv env baseValOpt (ty, overrides) =
       Info=UnknownValue}
 
 //-------------------------------------------------------------------------
-// Optimize/analyze an application of an intrinsic operator to arguments
+// Make and optimize String.Concat calls
 //------------------------------------------------------------------------- 
 
+and MakeOptimizedSystemStringConcatCall cenv env m args =
+    let rec optimizeArg e accArgs =
+        match e, accArgs with
+        | Expr.Op(TOp.ILCall(_, _, _, _, _, _, _, methRef, _, _, _), _, [ Expr.Op(TOp.Array, _, args, _) ], _), _ when IsSystemStringConcatArray methRef ->
+            optimizeArgs args accArgs
+
+        | Expr.Op(TOp.ILCall(_, _, _, _, _, _, _, methRef, _, _, _), _, args, _), _ when IsSystemStringConcatOverload methRef ->
+            optimizeArgs args accArgs
+
+        // Optimize string constants, e.g. "1" + "2" will turn into "12"
+        | Expr.Const(Const.String str1, _, _), Expr.Const(Const.String str2, _, _) :: accArgs ->
+            mkString cenv.g m (str1 + str2) :: accArgs
+
+        | arg, _ -> arg :: accArgs
+
+    and optimizeArgs args accArgs =
+        (args, accArgs)
+        ||> List.foldBack (fun arg accArgs -> optimizeArg arg accArgs)
+
+    let args = optimizeArgs args []
+
+    let expr' =
```

Comment:
Please don't use ticks. We try to get rid of them in the compiler since it's always an issue with tooling like debugger and immediate window and stuff

Reply:
Yea, agreed. `expr2` should be fine then, so we can debug it as well.